### PR TITLE
ARROW-5758: [C++][Gandiva][Java] Support casting decimals to varchar and vice versa

### PR DESCRIPTION
**Pull request by pprudhvi**
support `castVARCHAR(decimal, out_str_len)`, `castDECIMAL(string, out_precision, out_scale)` functions in Gandiva

### CODE REVIEW

File: cpp/src/gandiva/precompiled/decimal_wrapper.cc [hunk 1]
```diff
@@ -358,4 +358,36 @@ boolean is_distinct_from_decimal128_decimal128_internal(
...
+char* castVARCHAR_decimal128_int64_internal(int64_t context, int64_t x_high,
+                                            uint64_t x_low, int32_t x_precision,
+                                            int32_t x_scale, int64_t out_len_param,
+                                            int32_t* out_length) {
+  int32_t full_dec_str_len;
+  char* dec_str =
+      gdv_fn_dec_to_string(context, x_high, x_low, x_scale, &full_dec_str_len);
+  int32_t trunc_dec_str_len =
+      out_len_param < full_dec_str_len ? out_len_param : full_dec_str_len;
+  *out_length = trunc_dec_str_len;
```
Comment (pprudhvi) [inline cpp/src/gandiva/precompiled/decimal_wrapper.cc]: how do I free the remaining bytes if out_len < full_decimal_str_len?

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 2]
```diff
@@ -97,6 +97,26 @@ int32_t gdv_fn_populate_varlen_vector(int64_t context_ptr, int8_t* data_ptr,
...
   return 0;
 }
+
+void gdv_fn_dec_from_string(const char* in, int32_t in_length,
+                            int32_t* precision_from_str, int32_t* scale_from_str,
+                            int64_t* dec_high_from_str, uint64_t* dec_low_from_str) {
+  arrow::Decimal128 dec;
+  auto status = arrow::Decimal128::FromString(std::string(in, in_length), &dec,
+                                              precision_from_str, scale_from_str);
+  *dec_high_from_str = dec.high_bits();
```
Comment (praveenbingo) [inline cpp/src/gandiva/gdv_function_stubs.cc]: we seem to be ignoring status..if it is not ok then we should set error in context and return.
else this is likely to crash..

File: cpp/src/gandiva/precompiled/decimal_wrapper.cc [hunk 3]
```diff
@@ -358,4 +358,36 @@ boolean is_distinct_from_decimal128_decimal128_internal(
...
+void castDECIMAL_utf8_internal(const char* in, int32_t in_length, int32_t out_precision,
+                               int32_t out_scale, int64_t* out_high, uint64_t* out_low) {
+  int64_t dec_high_from_str;
+  uint64_t dec_low_from_str;
+  int32_t precision_from_str;
+  int32_t scale_from_str;
+  gdv_fn_dec_from_string(in, in_length, &precision_from_str, &scale_from_str,
+                         &dec_high_from_str, &dec_low_from_str);
+
+  gandiva::BasicDecimalScalar128 x({dec_high_from_str, dec_low_from_str},
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/decimal_wrapper.cc]: what happens when length is > 38..

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 4]
```diff
@@ -97,6 +97,26 @@ int32_t gdv_fn_populate_varlen_vector(int64_t context_ptr, int8_t* data_ptr,
...
+  *dec_low_from_str = dec.low_bits();
+}
+
+char* gdv_fn_dec_to_string(int64_t context, int64_t x_high, uint64_t x_low,
```
Comment (praveenbingo) [inline cpp/src/gandiva/gdv_function_stubs.cc]: good decision to split this to do only decimal to string! :)

File: cpp/src/gandiva/tests/decimal_test.cc [hunk 5]
```diff
@@ -843,4 +844,116 @@ TEST_F(TestDecimal, TestNullDecimalConstant) {
...
+  // Build a projector for the expressions.
+  std::shared_ptr<Projector> projector;
+
+  auto status = Projector::Make(schema, {expr}, TestConfiguration(), &projector);
+  EXPECT_TRUE(status.ok()) << status.message();
+
+  // Create a row-batch with some sample data
+  int num_records = 5;
+
+  auto array_str = MakeArrowArrayUtf8({"10.5134", "-0.0", "-0.1", "10.516", "-1000"},
```
Comment (praveenbingo) [inline cpp/src/gandiva/tests/decimal_test.cc]: please add non-valid numbers..

File: cpp/src/gandiva/tests/decimal_test.cc [hunk 6]
```diff
@@ -843,4 +844,116 @@ TEST_F(TestDecimal, TestNullDecimalConstant) {
...
+  std::shared_ptr<Projector> projector;
+
+  auto status = Projector::Make(schema, {expr, expr_1}, TestConfiguration(), &projector);
+  EXPECT_TRUE(status.ok()) << status.message();
+
+  // Create a row-batch with some sample data
+  int num_records = 5;
+  auto array_dec = MakeArrowArrayDecimal(
+      decimal_type,
+      MakeDecimalVector({"10.51", "1.23", "100.23", "-1000.23", "-0000.10"}, scale),
```
Comment (praveenbingo) [inline cpp/src/gandiva/tests/decimal_test.cc]: please add some thing greater than 38 digits with precision 38,0

File: cpp/src/gandiva/tests/decimal_test.cc [hunk 7]
```diff
@@ -843,4 +844,116 @@ TEST_F(TestDecimal, TestNullDecimalConstant) {
...
+  auto field_str = field("in_str", utf8());
+  auto schema = arrow::schema({field_str});
+
+  // output fields
+  auto res_dec = field("res_dec", decimal_type);
+
+  // build expressions.
+  auto node_str = TreeExprBuilder::MakeField(field_str);
+  auto cast_decimal =
+      TreeExprBuilder::MakeFunction("castDECIMAL", {node_str}, decimal_type);
```
Comment (praveenbingo) [inline cpp/src/gandiva/tests/decimal_test.cc]: please also add a wrapper test - 

1. convert decimal -> varchar
2. use varchar from 1 to convert to decimal..

just to ensure we do not crash when user pipes data back.

File: cpp/src/gandiva/gdv_function_stubs.cc [hunk 8]
```diff
@@ -97,6 +97,26 @@ int32_t gdv_fn_populate_varlen_vector(int64_t context_ptr, int8_t* data_ptr,
...
+                           int32_t x_scale, int32_t* dec_str_len) {
+  arrow::Decimal128 dec(arrow::BasicDecimal128(x_high, x_low));
+  std::string dec_str = dec.ToString(x_scale);
+  *dec_str_len = static_cast<int32_t>(dec_str.length());
+  char* ret = reinterpret_cast<char*>(gdv_fn_context_arena_malloc(context, *dec_str_len));
```
Comment (fsaintjacques) [inline cpp/src/gandiva/gdv_function_stubs.cc]: This function can return `NULL` on failure, you need to protect against this.

**Review by fsaintjacques — CHANGES_REQUESTED** [review]
See null comment.

File: cpp/src/gandiva/gdv_function_stubs.cc (hunk 8 above)
Comment (pprudhvi) [reply cpp/src/gandiva/gdv_function_stubs.cc]: now sets error in context and returns NULLPTR

**Review by praveenbingo — APPROVED** [review]
+1 Thanks @pprudhvi

**Comment by pravindra** [comment]
thanks @pprudhvi

**Pull request closed without merging**